cli/src/InterpreterApp.scala

```diff
@@ -0,0 +1,154 @@
+
+    opt[Boolean]('p', "parse")
+      .optional()
+      .action((f, c) => c.copy(parse = f))
+      .text("Parse the input file as text")
+
+    opt[String]('f', "filter")
+      .optional()
+      .action((f, c) => c.copy(traceFilter = f))
+      .text("Filter the traces. The parameter is a regular expression applied to the opcode, e.g.: 'mread|mwrite'")
+
+    opt[String]('l', "logfile-name")
+      .optional()
+      .action((f, c) => c.copy(tracePattern = f))
+      .text("SWAM uses a logger based on [[https://docs.oracle.com/en/java/javase/13/docs/api/java.logging/java/util/logging/package-summary.html java.util.logging]]. Provide any valid valid file pattern")
+
+    opt[Boolean]('d', "debug-compiler")
+      .optional()
+      .action((f, c) => c.copy(debugCompiler = f))
+      .text("Activates the debug option for thee text parser")
+
+    opt[Boolean]('t', "trace")
+      .optional()
+      .action((f, c) => c.copy(trace = f))
+      .text("Traces WASM execution channels; stack, memory and opcodes")
+
+  }
+
+  def buffer = {
+    val buffer = ByteBuffer.allocate(2 * pageSize)
+    buffer.limit(pageSize)
+    buffer
+  }
+
+  def imports() =
+    Imports[IO](
+      TCMap[String, AsIsIO](
+        "env" -> TCMap[String, AsIIO]("memory" -> buffer)
+      ))
+
+  def createInstance(blocker: Blocker, config: Config): IO[Instance[IO]] = {
+    for {
+      compiler <- Compiler[IO]
+      engine <- if (config.trace)
+        Engine[IO](
+          tracer = Option(JULTracer(config.traceOutDir, config.tracePattern, NoTimestampFormatter, config.traceFilter)))
+      else
+        Engine[IO]()
+
+      module <- if (config.parse)
+        engine.compile(compiler.stream(config.wasm.toPath, config.debugCompiler, blocker))
+      else
+        engine.compile(config.wasm.toPath, blocker, 4096)
+
+      instance <- engine.instantiate(module, imports())
+    } yield instance
+  }
+
+  def run(args: List[String]): IO[ExitCode] = parser.parse(args, Config()) match {
+    case Some(config) =>
+      Blocker[IO].use { blocker =>
+        for {
+          instance <- createInstance(blocker, config)
+          _ <- if (config.main.isEmpty) {
+            IO(println("Listing available functions..."))
+            instance.exports.list
+              .collect[String, FuncType] {
+                case (name, f: FuncType) =>
+                  (name, f)
+              }
+              .foreach {
+                case (name, tpe: FuncType) => {
+                  println(s"\t ${Console.GREEN} $name ${Console.RESET}${tpe.params.mkString("( ", ",", " )")} -> ${tpe.t
+                    .mkString(",")}")
+                }
+              }
+            IO()
+          } else {
+            for {
+              f <- instance.exports.function(config.main)
+              _ <- f.invoke(Vector(), None)
+            } yield IO(ExitCode.Success)
+          }
+        } yield ()
+
+        IO(ExitCode.Success)
```

Comment:
This is the only code that will get executed in this end. So your program is doing nothing.

the `for`-comprehension is syntactic sugar for `flatMap`s ending with a `map`, so when you write:

```scala
for {
  res1 <- something
  res2 <- somethingElse(res1)
} yield computeResult(res1, res2)
```

It is parsed as:

```scala
something.flatMap { res1 =>
  somethingElse(res1).map { res2 =>
    computeResult(res1, res2)
  }
}
```

The `yield` part is the `map` so must be pure.

In your case you can remove this line and simply return `ExitCode.Success` in the yield. The components of the `for` don't need to return anything at all (specially not the Success code).